File: project3.cpp
#include <stdint.h>
#include <iostream>
#include <iomanip>
#include <fstream>
#include <filesystem>
#include <string.h>
#include <vector>
#include <utility>
#define MAGIC "4348PRJ3"

struct Header
{
    char magic[8];
    uint64_t rootID;
    uint64_t nextBlockID;
};

class Node
{
public:
    uint64_t blockID;
    uint64_t parent_ID;
    uint64_t currentPairs;
    uint64_t keys[19];
    uint64_t values[19];
    uint64_t offsets[20];
};

int isBigEndian();
uint64_t reverseBytes(uint64_t x);
bool isIndexFile(std::string);
bool readHeader(std::string, Header &);
bool writeHeader(std::string, uint64_t, uint64_t);
Node loadNode(std::string, uint64_t);
bool saveNode(std::string, Node &);
Node createEmptyNode(uint64_t, uint64_t, bool);
bool isLeaf(Node &);
int findInsertPosition(Node &, uint64_t);
void splitChild(std::string, Node &, int, uint64_t &);
void insertNonFull(std::string, Node &, uint64_t, uint64_t, uint64_t &);
bool insertKey(std::string, uint64_t, uint64_t);
bool convertToUint64(const char *, uint64_t &);
bool searchKey(std::string, uint64_t, uint64_t &);

Node possibleNodes[3];
uint64_t rootID;
uint64_t nextBlockID;

int main(int argc, char **argv)
{
    if (isBigEndian() == 0)
    {
        rootID = reverseBytes(0);
        nextBlockID = reverseBytes(1);
    }
    else
    {
        rootID = 0;
        nextBlockID = 1;
    }

    std::fstream file;

    if (argc < 3)
    {
        std::cerr << "Usage: " << argv[0] << " <command> <index-file> <optional-params>" << std::endl;
        return 1;
    }
    std::string command = argv[1];
    std::string indexFile = argv[2];

    for (char &c : command)
    {
        c = tolower(c);
    }

    if (command == "insert")
    {
        uint64_t key;
        uint64_t value;
        if (indexFile.empty())
        {
            std::cerr << "Usage: " << command << " insert <index-file> <key> <value>" << std::endl;
            return 1;
        }
        if (!isIndexFile(indexFile))
        {
            std::cerr << "Invalid file, file must be an index file";
            return 1;
        }
        if (convertToUint64(argv[3], key) && convertToUint64(argv[4], value))
        {
            std::cout << "Inserting...." << std::endl;

            insertKey(indexFile, key, value);
        }
        else
        {
            std::cerr << "Invalid key" << std::endl;
        }
    }
    else if (command == "create")
    {
        if (indexFile.empty())
        {
            std::cerr << "Usage: " << command << " insert <index-file> <key> <value>" << std::endl;
            return 1;
        }
        std::string fileToCreate = indexFile + ".idx";

        if (std::filesystem::exists(fileToCreate))
        {
            std::cerr << "File already exists" << std::endl;
            return 1;
        }

        file.open(fileToCreate, std::ios::out | std::ios::binary);
        char zeros[488] = {0};
        if (file.is_open())
        {
            file.write(MAGIC, 8);
            file.write((char *)&rootID, 8);
            file.write((char *)&nextBlockID, 8);
            file.write(zeros, 488);
            file.close();
            std::cout << "File Created Successfully " << std::endl;
        }
        else
        {
            std::cerr << "Could not create file" << std::endl;
            return 1;
        }
    }
    else if (command == "search")
    {
        uint64_t key;
        uint64_t value;
        if (indexFile.empty())
        {
            std::cerr << "Usage: " << command << " search <index-file> <key>" << std::endl;
            return 1;
        }
        if (!isIndexFile(indexFile))
        {
            std::cerr << "Invalid file, file must be an index file";
            return 1;
        }

        if (convertToUint64(argv[3], key))
        {
            std::cout << "Searching...." << std::endl;

            if (!searchKey(indexFile, key, value))
            {
                std::cout << "Not found" << std::endl;
            }
            else
            {
                std::cout << key << "," << value << std::endl;
            }
        }
    }
    else if (command == "load")
    {
        std::string csvFile = argv[3];
        if (indexFile.empty() || csvFile.empty())
        {
            std::cerr << "Usage: " << command << " load <index-file> <csv-file>" << std::endl;
            return 1;
        }
        if (!isIndexFile(indexFile))
        {
            std::cerr << "Invalid file, file must be an index file";
            return 1;
        }
    }
    else if (command == "print")
    {
        if (indexFile.empty())
        {
            std::cerr << "Usage: " << command << " insert <index-file> <key> <value>" << std::endl;
            return 1;
        }
        if (!isIndexFile(indexFile))
        {
            std::cerr << "Invalid file, file must be an index file";
            return 1;
        }
    }
    else if (command == "extract")
    {
        std::string csvFile = argv[3];
        if (indexFile.empty() || csvFile.empty())
        {
            std::cerr << "Usage: " << command << " load <index-file> <csv-file>" << std::endl;
            return 1;
        }
        if (!isIndexFile(indexFile))
        {
            std::cerr << "Invalid file, file must be an index file";
            return 1;
        }
    }
    else
    {
        std::cerr << "Usage: " << argv[0] << " <command> <index-file> <optional-params>" << std::endl;
        return 1;
    }
    return 0;
}

int isBigEndian()
{
    int num = 1;
    return (*(char *)&num == 0);
}

uint64_t reverseBytes(uint64_t x)
{
    uint8_t dest[sizeof(uint64_t)];
    uint8_t *src = (uint8_t *)&x;
    for (int c = 0; c < sizeof(uint64_t); c++)
    {
        dest[c] = src[sizeof(uint64_t) - c - 1];
    }

    return *(uint64_t *)dest;
}

bool isIndexFile(std::string filename)
{
    std::ifstream infile(filename, std::ios::binary);
    if (!infile)
    {
        std::cerr << "File does not exist, please create it first" << std::endl;
        return false;
    }
    char buffer[8];

    infile.read(buffer, 8);

    if (strncmp(buffer, MAGIC, 8) == 0)
    {
        return true;
    }

    return false;
}

bool readHeader(std::string filename, Header &header)
{
    std::ifstream file(filename, std::ios::binary);
    if (!file)
    {
        return false;
    }

    file.read(header.magic, 8);

    file.read((char *)&header.rootID, 8);
    if (!isBigEndian())
    {
        header.rootID = reverseBytes(header.rootID);
    }

    file.read((char *)&header.nextBlockID, 8);
    if (!isBigEndian())
    {
        header.nextBlockID = reverseBytes(header.nextBlockID);
    }

    file.close();
    return true;
}

bool writeHeader(std::string filename, uint64_t rootID, uint64_t nextBlockID)
{
    std::fstream file(filename, std::ios::in | std::ios::out | std::ios::binary);
    if (!file)
    {
        return false;
    }

    uint64_t root = rootID;
    uint64_t next = nextBlockID;
    if (!isBigEndian())
    {
        root = reverseBytes(root);
        next = reverseBytes(next);
    }

    file.seekp(0, std::ios::beg);
    file.write(MAGIC, 8);
    file.write((char *)&root, 8);
    file.write((char *)&next, 8);

    file.close();
    return true;
}

Node loadNode(std::string filename, uint64_t blockID)
{
    Node node;
    std::ifstream file(filename, std::ios::binary);

    file.seekg(blockID * 512, std::ios::beg);

    file.read((char *)&node.blockID, 8);
    if (!isBigEndian())
    {
        node.blockID = reverseBytes(node.blockID);
    }

    file.read((char *)&node.parent_ID, 8);
    if (!isBigEndian())
    {
        node.parent_ID = reverseBytes(node.parent_ID);
    }

    file.read((char *)&node.currentPairs, 8);
    if (!isBigEndian())
    {
        node.currentPairs = reverseBytes(node.currentPairs);
    }

    for (int i = 0; i < 19; i++)
    {
        file.read((char *)&node.keys[i], 8);
        if (!isBigEndian())
        {
            node.keys[i] = reverseBytes(node.keys[i]);
        }
    }

    for (int i = 0; i < 19; i++)
    {
        file.read((char *)&node.values[i], 8);
        if (!isBigEndian())
        {
            node.values[i] = reverseBytes(node.values[i]);
        }
    }

    for (int i = 0; i < 20; i++)
    {
        file.read((char *)&node.offsets[i], 8);
        if (!isBigEndian())
        {
            node.offsets[i] = reverseBytes(node.offsets[i]);
        }
    }

    file.close();
    return node;
}

bool saveNode(std::string filename, Node &node)
{
    std::fstream file(filename, std::ios::in | std::ios::out | std::ios::binary);
    if (!file)
    {
        return false;
    }

    file.seekp(node.blockID * 512, std::ios::beg);

    uint64_t blockID = node.blockID;
    if (!isBigEndian())
    {
        blockID = reverseBytes(blockID);
    }
    file.write((char *)&blockID, 8);

    uint64_t parentID = node.parent_ID;
    if (!isBigEndian())
    {
        parentID = reverseBytes(parentID);
    }
    file.write((char *)&parentID, 8);

    uint64_t numKeys = node.currentPairs;
    if (!isBigEndian())
    {
        numKeys = reverseBytes(numKeys);
    }
    file.write((char *)&numKeys, 8);

    for (int i = 0; i < 19; i++)
    {
        uint64_t key = node.keys[i];
        if (!isBigEndian())
        {
            key = reverseBytes(key);
        }
        file.write((char *)&key, 8);
    }

    for (int i = 0; i < 19; i++)
    {
        uint64_t value = node.values[i];
        if (!isBigEndian())
        {
            value = reverseBytes(value);
        }
        file.write((char *)&value, 8);
    }

    for (int i = 0; i < 20; i++)
    {
        uint64_t offset = node.offsets[i];
        if (!isBigEndian())
        {
            offset = reverseBytes(offset);
        }
        file.write((char *)&offset, 8);
    }

    char zeros[8] = {0};
    file.write(zeros, 8);

    file.close();
    return true;
}

Node createEmptyNode(uint64_t blockID, uint64_t parentID, bool isLeaf)
{
    Node node;
    node.blockID = blockID;
    node.parent_ID = parentID;
    node.currentPairs = 0;

    for (int i = 0; i < 19; i++)
    {
        node.keys[i] = 0;
        node.values[i] = 0;
    }

    for (int i = 0; i < 20; i++)
    {
        node.offsets[i] = 0;
    }

    return node;
}

bool isLeaf(Node &node)
{
    for (int i = 0; i < 20; i++)
    {
        if (node.offsets[i] != 0)
        {
            return false;
        }
    }
    return true;
}

int findInsertPosition(Node &node, uint64_t key)
{
    int pos = 0;
    while (pos < node.currentPairs && key > node.keys[pos])
    {
        pos++;
    }
    return pos;
}

void splitChild(std::string filename, Node &parent, int childIndex, uint64_t &nextBlockID)
{

    Node fullChild = loadNode(filename, parent.offsets[childIndex]);

    Node newNode = createEmptyNode(nextBlockID, parent.blockID, isLeaf(fullChild));
    nextBlockID++;

    int midIndex = 9;

    newNode.currentPairs = 9;
    for (int i = 0; i < 9; i++)
    {
        newNode.keys[i] = fullChild.keys[midIndex + 1 + i];
        newNode.values[i] = fullChild.values[midIndex + 1 + i];
    }

    if (!isLeaf(fullChild))
    {
        for (int i = 0; i < 10; i++)
        {
            newNode.offsets[i] = fullChild.offsets[midIndex + 1 + i];
        }

        for (int i = 0; i < 10; i++)
        {
            if (newNode.offsets[i] != 0)
            {
                Node child = loadNode(filename, newNode.offsets[i]);
                child.parent_ID = newNode.blockID;
                saveNode(filename, child);
            }
        }
    }

    fullChild.currentPairs = 9;

    for (int i = 9; i < 19; i++)
    {
        fullChild.keys[i] = 0;
        fullChild.values[i] = 0;
    }
    if (!isLeaf(fullChild))
    {
        for (int i = 10; i < 20; i++)
        {
            fullChild.offsets[i] = 0;
        }
    }

    for (int i = parent.currentPairs; i > childIndex; i--)
    {
        parent.keys[i] = parent.keys[i - 1];
        parent.values[i] = parent.values[i - 1];
        parent.offsets[i + 1] = parent.offsets[i];
    }

    parent.keys[childIndex] = fullChild.keys[midIndex];
    parent.values[childIndex] = fullChild.values[midIndex];
    parent.offsets[childIndex + 1] = newNode.blockID;
    parent.currentPairs++;

    saveNode(filename, fullChild);
    saveNode(filename, newNode);
    saveNode(filename, parent);
}

void insertNonFull(std::string filename, Node &node, uint64_t key, uint64_t value, uint64_t &nextBlockID)
{
    int i = node.currentPairs - 1;

    if (isLeaf(node))
    {
        while (i >= 0 && key < node.keys[i])
        {
            node.keys[i + 1] = node.keys[i];
            node.values[i + 1] = node.values[i];
            i--;
        }

        node.keys[i + 1] = key;
        node.values[i + 1] = value;
        node.currentPairs++;

        saveNode(filename, node);
    }
    else
    {
        while (i >= 0 && key < node.keys[i])
        {
            i--;
        }
        i++;

        Node child = loadNode(filename, node.offsets[i]);

        if (child.currentPairs == 19)
        {
            splitChild(filename, node, i, nextBlockID);

            node = loadNode(filename, node.blockID);

            if (key > node.keys[i])
            {
                i++;
            }

            child = loadNode(filename, node.offsets[i]);
        }

        insertNonFull(filename, child, key, value, nextBlockID);
    }
}

bool insertKey(std::string filename, uint64_t key, uint64_t value)
{
    Header header;
    if (!readHeader(filename, header))
    {
        return false;
    }

    uint64_t rootID = header.rootID;
    uint64_t nextBlockID = header.nextBlockID;

    if (rootID == 0)
    {

        Node root = createEmptyNode(nextBlockID, 0, true);
        root.keys[0] = key;
        root.values[0] = value;
        root.currentPairs = 1;

        saveNode(filename, root);

        rootID = nextBlockID;
        nextBlockID++;
        writeHeader(filename, rootID, nextBlockID);

        return true;
    }

    Node root = loadNode(filename, rootID);

    if (root.currentPairs == 19)
    {
        Node newRoot = createEmptyNode(nextBlockID, 0, false);
        nextBlockID++;

        newRoot.offsets[0] = root.blockID;
        root.parent_ID = newRoot.blockID;
        saveNode(filename, root);

        splitChild(filename, newRoot, 0, nextBlockID);

        rootID = newRoot.blockID;
        writeHeader(filename, rootID, nextBlockID);

        newRoot = loadNode(filename, newRoot.blockID);
        insertNonFull(filename, newRoot, key, value, nextBlockID);

        writeHeader(filename, rootID, nextBlockID);
    }
    else
    {
        insertNonFull(filename, root, key, value, nextBlockID);

        writeHeader(filename, rootID, nextBlockID);
    }

    return true;
}

bool convertToUint64(const char *str, uint64_t &result)
{
    if (str == nullptr)
    {
        return false;
    }

    try
    {
        result = std::stoull(str);
        return true;
    }
    catch (...)
    {
        return false;
    }
}

bool searchKey(std::string filename, uint64_t key, uint64_t &resultValue)
{
    Header header;
    if (!readHeader(filename, header))
    {
        return false;
    }

    if (header.rootID == 0)
    {
        return false;
    }

    uint64_t currentBlockID = header.rootID;

    while (currentBlockID != 0)
    {
        Node node = loadNode(filename, currentBlockID);

        int i = 0;
        while (i < node.currentPairs && key > node.keys[i])
        {
            i++;
        }

        if (i < node.currentPairs && key == node.keys[i])
        {
            resultValue = node.values[i];
            return true;
        }

        if (isLeaf(node))
        {
            return false;
        }

        currentBlockID = node.offsets[i];
    }

    return false;
}

File: project3.py
import os
import sys
import struct

# node cache to enforce 3 node limit
node_cache = {}
cache_order = [] #track order for lru algo

def get_cached_node(file, block_id):
    # check if node is already in cache and move to end (recently used)
    if block_id in node_cache:
        cache_order.remove(block_id)
        cache_order.append(block_id)
        return node_cache[block_id]
    
    node = read_node_at_block(file, block_id)
    
    # if cache is full, evict least recently used
    if len(node_cache) >= 3:
        lru_block_id = cache_order.pop(0)
        del node_cache[lru_block_id]
    
    #add node
    node_cache[block_id] = node
    cache_order.append(block_id)
    
    return node

def clear_cache():
    global node_cache, cache_order
    node_cache = {}
    cache_order = []

def create(filename):
    # check if file already exists
    if os.path.exists(filename):
        print("Error: File already exists")
        return False

    try:
        with open(filename, "wb") as file:
            # write header: magic number, root id (0 for empty tree), next block id (1)
            MAGIC_NUMBER = b"4348PRJ3"
            root_block_id = (0).to_bytes(8, "big")
            next_block_id = (1).to_bytes(8, "big")
            unused_space = bytes(512 - 24)
            file.write(MAGIC_NUMBER + root_block_id + next_block_id + unused_space)
        return True
    except Exception as e:
        print(f"Error creating file: {e}")
        return False

def insert(filename, key, value):
    try:
        clear_cache()  
        with open(filename, "rb+") as file:
            # read header to get root and next block ids
            file.seek(0)
            header = file.read(24)
            magic_number, root_block_id, next_block_id = struct.unpack(">8sQQ", header)

            # case 1: empty tree, create first node as root
            if root_block_id == 0:
                root_block_id = next_block_id
                next_block_id += 1

                root_node = create_btree_node(root_block_id, 0, [(key, value)])
                file.seek(root_block_id * 512)
                file.write(root_node)

                # update header with new root
                file.seek(8)
                file.write(struct.pack(">QQ", root_block_id, next_block_id))
                return True

            root = get_cached_node(file, root_block_id)
            
            # case 2: root is full, need to split
            if root["num_keys"] == 19:
                new_root_block_id = next_block_id
                next_block_id += 1
                
                # split the old root
                next_block_id, middle_key, middle_value, new_node_block_id = split_node(
                    file, root_block_id, next_block_id, root_block_id
                )
                
                # create new root with middle key
                new_root_keys = [middle_key] + [0] * 18
                new_root_values = [middle_value] + [0] * 18
                new_root_children = [root_block_id, new_node_block_id] + [0] * 18 # old root and new node as children
                
                node_format = ">QQQ" + "Q" * 19 + "Q" * 19 + "Q" * 20
                new_root_node = struct.pack(node_format,
                    new_root_block_id, 0, 1,
                    *new_root_keys, *new_root_values, *new_root_children)
                
                file.seek(new_root_block_id * 512)
                file.write(new_root_node)
                
                # update parent ids of children
                update_parent_id(file, root_block_id, new_root_block_id)
                update_parent_id(file, new_node_block_id, new_root_block_id)
                
                # update header with new root
                file.seek(8)
                file.write(struct.pack(">QQ", new_root_block_id, next_block_id))
                
                root_block_id = new_root_block_id
            
            # insert into non-full root
            next_block_id = insert_non_full(file, root_block_id, key, value, next_block_id)
            
            # update next block id in header
            file.seek(16)
            file.write(struct.pack(">Q", next_block_id))
            
            return True
    except Exception as e:
        print(f"Error during insert: {e}")
        return False
    finally:
        clear_cache() 

def insert_non_full(file, block_id, key, value, next_block_id):
    node = get_cached_node(file, block_id)
    
    # case 1: node is a leaf, insert directly
    if is_leaf(node):
        keys = node["keys"][:node["num_keys"]]
        values = node["values"][:node["num_keys"]]
        
        # find position to insert key
        insert_pos = 0
        while insert_pos < len(keys) and keys[insert_pos] < key:
            insert_pos += 1
        
        # insert key and value at position
        keys.insert(insert_pos, key)
        values.insert(insert_pos, value)
        
        keys_padded = keys + [0] * (19 - len(keys))
        values_padded = values + [0] * (19 - len(values))
        
        # write updated node back to disk
        node_format = ">QQQ" + "Q" * 19 + "Q" * 19 + "Q" * 20
        updated_node = struct.pack(node_format,
            node["block_id"], node["parent_id"], len(keys),
            *keys_padded, *values_padded, *node["children"])
        
        file.seek(block_id * 512)
        file.write(updated_node)
        
        return next_block_id
    else:
        # case 2: internal node, find correct child
        i = 0
        while i < node["num_keys"] and key > node["keys"][i]:
            i += 1
        
        child_id = node["children"][i]
        child = get_cached_node(file, child_id)
        
        # if child is full, split it first
        if child["num_keys"] == 19:
            next_block_id, middle_key, middle_value, new_child_id = split_node(
                file, child_id, next_block_id, block_id
            )
            
            # reload parent after split
            node = get_cached_node(file, block_id)
            
            # insert middle key into parent
            keys = list(node["keys"][:node["num_keys"]])
            values = list(node["values"][:node["num_keys"]])
            children = list(node["children"][:node["num_keys"] + 1])
            
            insert_pos = 0
            while insert_pos < len(keys) and keys[insert_pos] < middle_key:
                insert_pos += 1
            
            keys.insert(insert_pos, middle_key)
            values.insert(insert_pos, middle_value)
            children.insert(insert_pos + 1, new_child_id)
            
            keys_padded = keys + [0] * (19 - len(keys))
            values_padded = values + [0] * (19 - len(values))
            children_padded = children + [0] * (20 - len(children))
            
            # write updated parent back
            node_format = ">QQQ" + "Q" * 19 + "Q" * 19 + "Q" * 20
            updated_node = struct.pack(node_format,
                node["block_id"], node["parent_id"], len(keys),
                *keys_padded, *values_padded, *children_padded)
            
            file.seek(block_id * 512)
            file.write(updated_node)
            
            # adjust child index if key is greater than middle key
            if key > middle_key:
                i += 1
            
            child_id = children[i]
        
        # recursively insert into child
        next_block_id = insert_non_full(file, child_id, key, value, next_block_id)
        
        return next_block_id

def split_node(file, block_id, next_block_id, parent_id):
    node = get_cached_node(file, block_id)
    
    # split at middle index (9)
    middle_index = 9
    middle_key = node["keys"][middle_index]
    middle_value = node["values"][middle_index]
    
    # left node gets keys 0-8
    left_keys = node["keys"][:middle_index] + [0] * (19 - middle_index)
    left_values = node["values"][:middle_index] + [0] * (19 - middle_index)
    left_children = node["children"][:middle_index + 1] + [0] * (20 - middle_index - 1)
    
    # right node gets keys 10-18
    right_keys = node["keys"][middle_index + 1:node["num_keys"]] + [0] * (19 - (node["num_keys"] - middle_index - 1))
    right_values = node["values"][middle_index + 1:node["num_keys"]] + [0] * (19 - (node["num_keys"] - middle_index - 1))
    right_children = node["children"][middle_index + 1:node["num_keys"] + 1] + [0] * (20 - (node["num_keys"] - middle_index))
    
    node_format = ">QQQ" + "Q" * 19 + "Q" * 19 + "Q" * 20
    
    # write left node at original block
    left_node = struct.pack(node_format,
        block_id, parent_id, middle_index,
        *left_keys, *left_values, *left_children)
    
    # write right node at new block
    new_block_id = next_block_id
    next_block_id += 1
    
    right_node = struct.pack(node_format,
        new_block_id, parent_id, node["num_keys"] - middle_index - 1,
        *right_keys, *right_values, *right_children)
    
    file.seek(block_id * 512)
    file.write(left_node)
    
    file.seek(new_block_id * 512)
    file.write(right_node)
    
    # update parent ids of children
    for child_id in left_children:
        if child_id != 0:
            update_parent_id(file, child_id, block_id)
    
    for child_id in right_children:
        if child_id != 0:
            update_parent_id(file, child_id, new_block_id)
    
    return next_block_id, middle_key, middle_value, new_block_id

def read_node_at_block(file, block_id):
    # seek to block position and read 512 bytes
    file.seek(block_id * 512)
    node_data = file.read(512)
    return read_btree_node(node_data)

def read_btree_node(data):
    # unpack node data from binary format
    node_data = data[:488] # skip unused bytes
    node_format = ">QQQ" + "Q" * 19 + "Q" * 19 + "Q" * 20
    unpacked = struct.unpack(node_format, node_data)

    return {
        "block_id": unpacked[0],
        "parent_id": unpacked[1],
        "num_keys": unpacked[2],
        "keys": list(unpacked[3:22]),
        "values": list(unpacked[22:41]),
        "children": list(unpacked[41:61]),
    }

def print_btree(filename):
    try:
        clear_cache()
        with open(filename, "rb") as file:
            file.seek(8)
            root_block_id = int.from_bytes(file.read(8), "big")
            
            if root_block_id == 0:
                return
            
            # collect all key-value pairs using dfs
            key_value_pairs = []
            dfs(file, root_block_id, key_value_pairs)
            
            # print each pair
            for key, value in key_value_pairs:
                print(f"{key},{value}")
    except Exception as e:
        print(f"Error printing: {e}")
    finally:
        clear_cache()

def dfs(file, block_id, key_value_pairs):
    # in-order traversal of b-tree
    if block_id == 0:
        return
    
    node = get_cached_node(file, block_id)
    
    for i in range(node["num_keys"]):
        dfs(file, node["children"][i], key_value_pairs) # visit left child
        key_value_pairs.append((node['keys'][i], node['values'][i])) # visit key
    
    dfs(file, node["children"][node["num_keys"]], key_value_pairs) # visit rightmost child

def extract(filename, output_file):
    if os.path.exists(output_file):
        print("Error: Output file already exists")
        return False

    try:
        clear_cache() 
        with open(filename, "rb") as file:
            file.seek(8)
            root_block_id = int.from_bytes(file.read(8), "big")

            if root_block_id == 0:
                return False

            # collect all pairs using dfs
            key_value_pairs = []
            dfs(file, root_block_id, key_value_pairs)

        # write to csv file
        with open(output_file, "w") as out_file:
            out_file.write("\n".join(f"{key},{value}" for key, value in key_value_pairs))
        return True
    except Exception as e:
        print(f"Error during extract: {e}")
        return False
    finally:
        clear_cache()

def load(filename, csv_file):
    if not os.path.exists(csv_file):
        print(f"Error: CSV file does not exist")
        return False

    try:
        # read csv and insert each pair
        with open(csv_file, "r") as file:
            for line in file:
                if not line.strip():
                    continue

                parts = line.strip().split(',')
                key = int(parts[0])
                value = int(parts[1])
                insert(filename, key, value)
        return True
    except Exception as e:
        print(f"Error during load: {e}")
        return False

def search(filename, key):
    try:
        clear_cache() 
        with open(filename, "rb") as file:
            file.seek(8)
            root_block_id = int.from_bytes(file.read(8), "big")

            if root_block_id == 0:
                return None

            # traverse tree to find key
            current_block_id = root_block_id

            while current_block_id != 0:
                node = get_cached_node(file, current_block_id)

                # check if key is in current node
                for i in range(node["num_keys"]):
                    if node["keys"][i] == key:
                        return node["values"][i]

                # find correct child to descend into
                i = 0
                while i < node["num_keys"] and key > node["keys"][i]:
                    i += 1

                current_block_id = node["children"][i]

            return None
    except Exception as e:
        print(f"Error during search: {e}")
        return None
    finally:
        clear_cache()

def update_parent_id(file, block_id, new_parent_id):
    # update parent id field in node
    file.seek(block_id * 512 + 8) # parent id is at offset 8
    file.write(struct.pack(">Q", new_parent_id))

def is_leaf(node):
    # node is leaf if all children are 0
    return all(child == 0 for child in node["children"])

def create_btree_node(block_id, parent_id, key_value_pairs):
    keys = [0] * 19
    for i, (key, _) in enumerate(key_value_pairs):
        keys[i] = key
    
    values = [0] * 19
    for i, (_, value) in enumerate(key_value_pairs):
        values[i] = value
    
    offset = [0] * 20
    
    # pack node into binary format
    node_format = ">QQQ" + "Q" * 19 + "Q" * 19 + "Q" * 20
    node_data = struct.pack(node_format, 
        block_id, parent_id, len(key_value_pairs), 
        *keys, *values, *offset)
    
    return node_data

def validate_index_file(filename):
    if not os.path.exists(filename):
        print("Error: File does not exist")
        return False
    
    try:
        with open(filename, "rb") as file:
            magic_number = file.read(8)
            if magic_number != b"4348PRJ3":
                print("Error: Invalid index file")
                return False
        return True
    except Exception as e:
        print(f"Error validating file: {e}")
        return False

def main():
    if len(sys.argv) < 3:
        print("Usage: python main.py <command> <index-file> <params>")
        return

    command = sys.argv[1].lower()
    index_file = sys.argv[2]

    if command == "create":
        create(index_file)
    elif command == "insert":
        if len(sys.argv) < 5:
            print("Usage: insert <index-file> <key> <value>")
            return
        if not validate_index_file(index_file):
            return
        key = int(sys.argv[3])
        value = int(sys.argv[4])
        insert(index_file, key, value)
    elif command == "search":
        if len(sys.argv) < 4:
            print("Usage: search <index-file> <key>")
            return
        if not validate_index_file(index_file):
            return
        key = int(sys.argv[3])
        result = search(index_file, key)
        if result is not None:
            print(f"{key},{result}")
        else:
            print("Not found")
    elif command == "load":
        if len(sys.argv) < 4:
            print("Usage: load <index-file> <csv-file>")
            return
        if not validate_index_file(index_file):
            return
        csv_file = sys.argv[3]
        load(index_file, csv_file)
    elif command == "print":
        if not validate_index_file(index_file):
            return
        print_btree(index_file)
    elif command == "extract":
        if len(sys.argv) < 4:
            print("Usage: extract <index-file> <output-file>")
            return
        if not validate_index_file(index_file):
            return
        output_file = sys.argv[3]
        extract(index_file, output_file)

if __name__ == "__main__":
    main()